File: src/elixir/fault_supervisor.ex
defmodule SDACS.FaultSupervisor do
  @moduledoc """
  Phase 633: Fault Supervisor — Elixir
  SDACS fault detection and recovery supervisor for drone fleet.
  """
  use Supervisor

  def start_link(opts \\ []) do
    Supervisor.start_link(__MODULE__, opts, name: __MODULE__)
  end

  @impl true
  def init(_opts) do
    children = [
      {SDACS.FaultDetector, []},
      {SDACS.RecoveryManager, []},
    ]
    Supervisor.init(children, strategy: :one_for_one)
  end

  def report_fault(drone_id, fault_type, details) do
    SDACS.FaultDetector.handle_fault(drone_id, fault_type, details)
  end
end

defmodule SDACS.FaultDetector do
  use GenServer

  def start_link(_opts), do: GenServer.start_link(__MODULE__, %{}, name: __MODULE__)
  def handle_fault(drone_id, type, details),
    do: GenServer.cast(__MODULE__, {:fault, drone_id, type, details})

  @impl true
  def init(state), do: {:ok, Map.put(state, :faults, [])}

  @impl true
  def handle_cast({:fault, drone_id, type, details}, state) do
    fault = %{drone_id: drone_id, type: type, details: details, at: DateTime.utc_now()}
    {:noreply, update_in(state, [:faults], &[fault | &1])}
  end
end

defmodule SDACS.RecoveryManager do
  use GenServer

  def start_link(_opts), do: GenServer.start_link(__MODULE__, %{}, name: __MODULE__)

  @impl true
  def init(state), do: {:ok, Map.put(state, :recoveries, 0)}

  @impl true
  def handle_cast({:recover, _drone_id}, state) do
    {:noreply, update_in(state, [:recoveries], &(&1 + 1))}
  end
end

File: src/erlang/fault_supervisor.erl
%% Phase 575: Fault Supervisor — Erlang OTP
%% SDACS fault-tolerant drone supervision tree

-module(fault_supervisor).
-behaviour(supervisor).

-export([start_link/0, start_link/1, init/1]).
-export([add_drone/2, remove_drone/1, restart_drone/1, drone_status/1]).
-export([get_all_drones/0, supervisor_stats/0]).

%% Supervisor public API

start_link() ->
    start_link([]).

start_link(Args) ->
    supervisor:start_link({local, ?MODULE}, ?MODULE, Args).

%% Supervisor callback

init(_Args) ->
    SupFlags = #{
        strategy  => one_for_one,
        intensity => 5,
        period    => 10
    },
    DroneSpecs = default_drone_specs(),
    {ok, {SupFlags, DroneSpecs}}.

%% Dynamic drone management

add_drone(DroneId, DroneConfig) ->
    ChildSpec = drone_child_spec(DroneId, DroneConfig),
    supervisor:start_child(?MODULE, ChildSpec).

remove_drone(DroneId) ->
    supervisor:terminate_child(?MODULE, DroneId),
    supervisor:delete_child(?MODULE, DroneId).

restart_drone(DroneId) ->
    supervisor:restart_child(?MODULE, DroneId).

drone_status(DroneId) ->
    Children = supervisor:which_children(?MODULE),
    case lists:keyfind(DroneId, 1, Children) of
        {DroneId, Pid, worker, _} when is_pid(Pid) ->
            {ok, running, Pid};
        {DroneId, undefined, worker, _} ->
            {ok, stopped};
        {DroneId, restarting, worker, _} ->
            {ok, restarting};
        false ->
            {error, not_found}
    end.

get_all_drones() ->
    Children = supervisor:which_children(?MODULE),
    [{Id, Status} || {Id, Status, worker, _} <- Children].

supervisor_stats() ->
    Children = supervisor:which_children(?MODULE),
    Running = length([P || {_, P, _, _} <- Children, is_pid(P)]),
    Stopped = length([P || {_, P, _, _} <- Children, P =:= undefined]),
    #{
        total    => length(Children),
        running  => Running,
        stopped  => Stopped,
        module   => ?MODULE
    }.

%% Internal helpers

drone_child_spec(DroneId, Config) ->
    #{
        id       => DroneId,
        start    => {drone_worker, start_link, [DroneId, Config]},
        restart  => transient,
        shutdown => 5000,
        type     => worker,
        modules  => [drone_worker]
    }.

default_drone_specs() ->
    %% Default drone workers started at supervisor init
    [].
